src/core/NewDomain.ml

```diff
@@ -1787,9 +1789,8 @@ struct
             ConAbsFace.gen @@ fun s s' (el : con) ->
             let rel_ss' = Rel.equate' s s' rel in
             let elx = make_coe rel_ss' (`Atom x) r' ~abs:(ConAbs.run rel_ss' abs) @@ Val.make el in
-            (* TODO: Favonia, is this actually safe? Shouldn't I be freshening? Also I feel weird
-               about all this unleash/make.*)
-            Abs (x, elx)
+            let x', pi = Perm.freshen_name x in
+            Abs (x', Con.swap pi elx)
```

Comment:
This is unfortunately wrong because `el` might have access to the outside `x` which was then shadowed. I wanted to fix the code but am not sure which part can depend on the inner, bound `x`. Can `data` and `genv` depend on `x`, for example?

Reply:
`data` can depend, but `genv` cannot.

Reply:
So `constr`, `constrs`, `data`, `tboundary`, `env_ps`, `sys` and `intro` are all morally under the binder `x`, and `genv` refers to the global env so it cannot depend on `x` in any way? Did I miss something?

I am worried about the innermost `benv` which extends `env_ps`. Should we apply `subst r x` to `env_ps`? Anyways I am not sure which part is under the binder `x`. If something only makes sense under the `x` binder, `Rel.hide' x rel` needs to be used instead of `x`.

Reply:
It will help me a lot if you could rename or mark the variables in some systematic way with the dependency and shadowing.

Reply:
I'm a little puzzled by all the shadowing issues. Shouldn't we just be freshening when we unbind???

Reply:
Maybe we should, and then that will be at right after `let Abs (x, data_tyx) = data_abs in`.

Reply:
I feel that any time we pattern match on `Abs (x, t)`, we are probably introducing a bug. It's most likely impossible to perfect the invariants generally so that this is safe.

We should have some "unleash" function or something, like we used to, and use this all the time. Then, in specific cases where we know what we are doing *locally*, we can do something more efficient, like doing the unsafe match.

Does this make sense to you?

Reply:
Mostly agreed, except that I want to give the carefully crafted V and fhcom types a try.

Reply:
OK cool! Let's have a safe API available so that low-dimensions people like me can use it, but allow people who have grasped the Geometry to use `Abs` directly.